### DISCUSSION TITLE
Using RCTEventEmitter with the new architecture

### DISCUSSION THREAD

**Discussion by BrandonMA** [reactwg/react-native-new-architecture#193] [Q&A]
We have a legacy application at my company that uses a custom method for loading React. We aim to follow the RCTAppDelegate closely but have made some adjustments to fit our needs, such as not directly inheriting from a custom AppDelegate. So far, the new architecture is working well. However, we encounter an issue when building a TurboModule using RCTEventEmitter on iOS. The module is intended to handle custom navigation, and while the `BackHandler.addEventListener` works fine on Android, it fails on iOS. All other TurboModules function correctly.

The specific iOS error is:
```
Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'Error when sending event: onBackPress with body: (null). RCTCallableJSModules is not set. This is probably because you've explicitly synthesized the RCTCallableJSModules in NavigationModule, even though it's inherited from RCTEventEmitter.'
```

Below is a simplified example of our code:

**NavigationModule.h**
```objc
#import <React/RCTEventEmitter.h>
#import "NavigationModuleImpl.h"
#import "NavigationDelegateProtocol.h"

@interface NavigationModule : RCTEventEmitter

@property (weak, nonatomic) NSObject<NavigationDelegateProtocol>* navigationDelegate;

+ (NavigationModule *)shared;
- (void)sendBackEvent;

@end
```

**NavigationModule.mm**
```objc
#import <Foundation/Foundation.h>
#import "RNNavigationModule.h"
#import "RNNavigationModuleSpec.h"
#import <React/RCTEventEmitter.h>

@interface NavigationModule () <NativeNavigationModuleSpec>
@end

@implementation NavigationModule {
    BOOL hasListeners;
}

- (instancetype)init {
    self = [super init];
    return self;
}

+ (BOOL)requiresMainQueueSetup {
    return NO;
}

RCT_EXPORT_MODULE()

static NavigationModule *_shared = nil;

+ (NavigationModule *)shared {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        if (_shared == nil) {
            _shared = [[self alloc] init];
        }
    });
    return _shared;
}

- (void)goBack {
    [[NavigationModule shared].navigationDelegate goBack];
}

- (void)navigate:(NSString *)featureName navigationType:(NSString *)navigationType params:(NSString *)params {
    [[NavigationModule shared].navigationDelegate navigate:featureName navigationType:navigationType params:params];
}

- (NSArray<NSString *> *)supportedEvents {
    return @[@"onBackPress"];
}

- (void)sendBackEvent {
    [self sendEventWithName:@"onBackPress" body:nil];
}

- (std::shared_ptr<facebook::react::TurboModule>)getTurboModule:
    (const facebook::react::ObjCTurboModule::InitParams &)params {
    return std::make_shared<facebook::react::NativeNavigationModuleSpecJSI>(params);
}

@end
```

The issue occurs when `sendBackEvent` is called on the native side. If you have any ideas on why this might be failing, I would greatly appreciate your input. 

Thank you for your help!

---

This version is more straightforward and breaks down the problem and code into easier-to-digest sections.

**Comment by BrandonMA** [comment]
In case anybody checks this, the problem was that I had two instances due to my use of a singleton, the singleton pattern works for other kind of turbo modules but not this, if you want to access the shared instance what's the recommended way to do so? So far I use some of the internal methods to get it but not sure what the "correct" way would be.

**Reply by shylugovind** [reply]
#ifdef RCT_NEW_ARCH_ENABLED

#import <LPMessagingSDK/LPMessagingSDK.h>
@protocol LPMessagingSDKdelegate;

// Promise typedefs (RCTPromiseResolveBlock / RCTPromiseRejectBlock)
#import <React/RCTBridgeModule.h>
#import <React/RCTEventEmitter.h>

// Generated Swift header - tries multiple possible module names
// The actual header name depends on the consuming app's module name
#if __has_include("Arcade-Swift.h")
  #import "Arcade-Swift.h"
#elif __has_include("xxxxx-Swift.h")
  #import "xxxxx-Swift.h"
#elif __has_include(<xxxxx/xxxxx-Swift.h>)
  #import <xxxxx/xxxxx-Swift.h>
#else
  // The consuming app must ensure Swift interop is enabled
  // and xxxxSwiftModule is accessible
  #warning "Could not find the generated Swift header for xxxx_module."
#endif

// React + Codegen
#import <React/RCTBridge.h>

// TurboModule header
#import "xxxxxTurboModule.h"

@implementation xxxxTurboModule {
  xxxxxSwiftModule *_swiftModule;
}

RCT_EXPORT_MODULE()

- (instancetype)initWithBridge:(RCTBridge *)bridge moduleName:(NSString *)moduleName {
  if ((self = [super init])) {
    _swiftModule = [[xxxxSwiftModule alloc] init];
  }
  return self;
}

I have been getting error error : ISO C++ forbids forward references to 'enum' types while building for ios, How can i resolve this, Any gudance will be a great help . Thanks.
